### PR TITLE
New query GetPoolDistr

### PR DESCRIPTION
**Pull request by newhoggy**
# Description

This introduce a new query GetPoolDistr which extracts the stake snapshot. This query is required to improve the CPU and memory efficiency of the query `leadership-schedule` command in the CLI.

# Checklist

- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Commits have useful messages
    - [ ] New tests are added if needed and existing tests are updated
    - [ ] If this branch changes Consensus and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../ouroboros-consensus/docs/interface-CHANGELOG.md)
    - [ ] If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [ ] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [ ] Reviewer requested

## `cardano-api` changes

There will be a new `GetPoolDistr` which uses a lot less CPU and memory:

```
GetPoolDistr
    :: SL.KeyHash 'SL.StakePool (EraCrypto era)
    -> BlockQuery (ShelleyBlock proto era)
                  (SL.PoolDistr (EraCrypto era))
```

This query takes a set of pool ids so it is possible to query multiple pools at once.

## `cardano-cli` changes

A existing command `query leadership-schedule` will be be functionally unchanged, but use much less CPU and memory.

### CODE REVIEW

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (nfrisby) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:300]: minor: extra newlines

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (nfrisby) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:199]: @JaredCorduan @lehins I think I asked a similar question about `PState` on a recent&similar PR, but this time it's about `PoolDistr`:

Do we want to expose `PoolDistr` in the query API? That's promising some stability in it's interface.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (nfrisby) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:282]: Is this `calculatePoolDistr` perhaps too expensive?

**Review by nfrisby — CHANGES_REQUESTED** [review]
I'm Requesting Changes, mostly to ensure that we have a good answer to whether `calculatePoolDistr` is prohibitively expensive.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (JaredCorduan) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:199]: I think it is okay, but I do not know for sure.

Our definition of "big" (ie the data structures that we know will be finding a home one day on disk) is "at least the size of the number of registered stake credentials". The `PoolDistr` is a consolidation of the per-stake credential stake distribution, which is itself not "big", but is derived from a "big" structure. I am currently expecting that we will keep the pool distribution in memory, and that this query will not be disrupted. But it is intimately related to a "big" data structure that we do know will be disrupted.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (JaredCorduan) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:282]: It might be, but I don't know that we have any clear requirements on what is too expensive. Note that `SL._pstakeSet` is one of the data structure that we anticipate being moved to disk.

Comment (nfrisby) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:282]: This is for `query leadership-schedule`, so I think we have a good amount of headroom -- I wouldn't expect SPOs etc to be spamming this command (though it's easy enough to do that accidentally within a script).

Comment (newhoggy) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:282]: Is there a way to make this cheaper if we're only interested in a subset of pool ids?

Comment (JaredCorduan) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:282]: great question! The following might be faster:

Make a new version of `calculatePoolDistr` with signature
```
calculatePoolDistr' :: SnapShot crypto -> Set (KeyHash 'Staking crypto) -> PoolDistr crypto
```
which is identical to `calculatePoolDistr` except that after looking up the credential in the delegation map (see [here](https://github.com/input-output-hk/cardano-ledger/blob/dd4b6e31a05e3c02c6f00f0d53cf20e3b8b74468/eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/NewEpoch.hs#L214)), it also looks up the pool ID in the set of pool IDs provided by the user (and discards the result if not found).

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (newhoggy) [inline ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:284]: Using the more efficient `calculatePoolDistr'` function.

**Comment by newhoggy** [comment]
I reverted back to use `calculatePoolDistr` because some refactoring in `cardono-ledger` blocks merging the PR otherwise.  I would like to merge this version and in a future PR switch to use `calculatePoolDistr'`

File: ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/NodeToClient/Version.hs:38]: :+1: `NodeToClientV_14` hasn't been yet released. `node-1.35.x` comes with `NodeToClientV_13` as the latest version.

File: ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs
Comment (kevinhammond) [reply ouroboros-consensus-shelley/src/Ouroboros/Consensus/Shelley/Ledger/Query.hs:339]: I'd look at use cases before optimising too much. I think the two likely cases are:

- all pools
- a specific pool (and then repeat for multi-pool operators)

The command is likely to be used occasionally rather than frequently, I expect

**Review by nfrisby — APPROVED** [review]
Marcin's commented that the NTC version was correct. So that's good.

The question of `calculatePoolDistr`'s performance remains unresolved, but I don't think it's a blocker for this PR. We'll need to assess it, though. @newhoggy can you pursue that with some measurements etc after merging this (or before, if that's convenient)?

**Comment by newhoggy** [comment]
Yep.  I will evaluate after merging.

There is a [follow up PR](https://github.com/input-output-hk/ouroboros-network/pull/3949) that improves the performance, but that's blocked on a refactoring PR.

**Comment by newhoggy** [comment]
I didn't notice much difference in CPU/memory usage of the new version of `cardano-node`.  I think it's because `query leadershipship-schedule` was already very expensive to start with.

**Pull request merged**
